browser/src/UI/components/CommandLine.tsx

```diff
@@ -1,63 +1,72 @@
 import * as React from "react"
 import { connect } from "react-redux"
 import styled from "styled-components"
+
 import { fadeInAndDown } from "./animations"
+import { boxShadow } from "./common"
 
 import * as State from "./../../Editor/NeovimEditor/NeovimEditorStore"
 
-const CommandLineBackground = styled.div`
-    position: absolute;
-    top: 0px;
-    left: 0px;
-    bottom: 0px;
-    right: 0px;
-    background-color: rgba(0, 0, 0, 0.25);
-    display: flex;
-    flex-direction: column;
-    align-items: center;
-`
-
 const CommandLineBox = styled.div`
     position: relative;
     margin-top: 16px;
     padding: 8px;
     width: 75%;
     max-width: 900px;
-    background-color: ${p => p.theme.background};
-    box-shadow: 0 4px 8px 2px rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
+    background-color: ${p => p.theme["menu.background"]};
+    ${boxShadow};
     animation: ${fadeInAndDown} 0.08s ease-in;
+    box-sizing: border-box;
+    overflow-wrap: break-word;
 `
-const CommandLineInput = styled.input`
+const CommandLineOutput = styled.div`
+    white-space: pre-wrap; /* CRUCIAL to render white-space correctly */
+    position: relative;
     border: 0px;
     background-color: rgba(0, 0, 0, 0.2);
     font-size: 1.1em;
     box-sizing: border-box;
     width: 100%;
+    height: auto;
     padding: 8px;
     outline: none;
     color: white;
 `
 
+const Cursor = styled.span`
+    background-color: white;
+    width: 2px;
+    position: absolute;
+    bottom: 6px;
+    height: 1.3em;
+`
+
 export interface ICommandLineRendererProps {
-    visible: boolean,
-    content: Array<[any, string]> | null
-    position: number,
+    visible: boolean
+    content: string
+    position: number
     firstchar: string
-    // level: number
+    level: number
+    prompt: string
 }
 
 interface State {
-    focused: boolean,
+    focused: boolean
+    waiting: boolean
 }
 
-class CommandLineRenderer extends React.PureComponent<ICommandLineRendererProps, State> {
+class CommandLine extends React.PureComponent<ICommandLineRendererProps, State> {
     public state = {
         focused: false,
+        waiting: true,
     }
+    private timer: any
     private _inputElement: HTMLInputElement
 
-    public handleChange(e: React.ChangeEvent<HTMLInputElement>) {
-        // UI.Actions.setCommandLinePosition(1, 1)
+    public componentDidMount() {
+        this.timer = setTimeout(() => {
```

Comment:
Just curious - why do we need this timeout here? It feels a bit slow to me when I press the `:` key (like there is some perceptible delay) - I think it may be due to this? 

I tried setting this to `0` and it seemed to work OK, which seemed to suggest it might not be necessary anymore - but it could also be the case that I missed the scenario in which it is needed 😄

Reply:
@bryphe it relates to a scenario where you have command line commands mapped and hit them in quick succession for example if you try `map <tab> :bn<CR>` and hit tab several times in a row with the timeout set to 0 the screen flickers as the component mounts very briefly. Its there to cover any/most (hopefully) of these scenarios.

Reply:
Whilst it does seem counter intuitive to try and hide it once I enabled it to test and tried using the functionality during a normal session I realised that as a vim user who has dozens of quick commandline mappings like `:w` or `:b#` etc. which I usually execute quickly and dont really think about having that quick flicker is really not a great experience though if you feel it detracts from the usability or have a suggestion other than the timeout like some sort of component mounting throttle I can remove it/try and implement something like that

Reply:
Oops, sorry I missed your other message. Yes, that quick flicker is painful - I didn't realize that it actually pops up through those command-line mappings. It's brutal especially if you use them frequently.

I think ideally, it would only show up via an explicit trigger, like when the user specifically presses `:`, `\`, `?`. In the case of mappings, I'm not sure it makes sense to even show it - I didn't expect that behavior. I wonder if it would make sense to log an issue against Neovim for that case?

Reply:
Technically in neovim the same thing happens if you watch the command line, the command you entered appears so all the events we're hooked into occur ☹️ so it replicates normal vim (subpar) behaviour.
As for keypresses that sounds good, though one scenario I can see causing an issue is in the case of key remaps as in my case I've rebound `:` to `;` so the command line wouldn't trigger and whilst I could just add that exception I dont know how many users would have done a similar thing and what their mappings might be for example '\' remapped to '|' or something else which is fairly unpredictable or we could have it be that oni requires the default bindings or maybe instead offer configs option to trigger the commandline if the user wants to remap this 🤔 .

Reply:
Though annoying given that it results in a delay a 80-100ms timeout seems to be enough unless a user literally held down a mapping for a good bit of time